File: tests/create_token_test.rb
# frozen_string_literal: true

require 'base64'
require 'fileutils'
require 'json'
require 'minitest/autorun'
require 'openssl'
require 'tmpdir'

require_relative '../src/scripts/create_token'

class FakeGitHubClient
  attr_reader :calls

  def initialize(repo_installation:, access_token:)
    @repo_installation = repo_installation
    @access_token = access_token
    @calls = []
  end

  def request(method, request_path, auth_token:, body: nil)
    @calls << {
      method: method,
      path: request_path,
      auth_token: auth_token,
      body: body
    }

    return @repo_installation if request_path == '/repos/octo-org/repo-a/installation'
    return @access_token if request_path.end_with?('/access_tokens')

    raise GitHubAppToken::GitHubApiError, 'not found'
  end
end

class CreateTokenTest < Minitest::Test
  def setup
    @private_key_pem = OpenSSL::PKey::RSA.generate(2048).to_pem
    @repo_installation = JSON.parse(File.read(File.join(__dir__, 'fixtures/responses/repo-installation.json')))
    @access_token = JSON.parse(File.read(File.join(__dir__, 'fixtures/responses/access-token.json')))
  end

  def test_normalizes_raw_escaped_newline_and_base64_private_keys
    expected = @private_key_pem.end_with?("\n") ? @private_key_pem : "#{@private_key_pem}\n"

    assert_equal expected, GitHubAppToken.normalize_private_key(@private_key_pem)
    assert_equal expected, GitHubAppToken.normalize_private_key(@private_key_pem.gsub("\n", '\\n'))
    assert_equal expected, GitHubAppToken.normalize_private_key(Base64.strict_encode64(@private_key_pem),
                                                               base64: true)
  end

  def test_uses_app_id_parameter_and_skips_lookup_when_installation_id_is_explicit
    runtime = temp_runtime
    client = fake_client

    GitHubAppToken.create_token(base_options(output_json: runtime[:output_json]),
                                env: env_with_private_key(runtime),
                                client: client,
                                now: Time.utc(2026, 5, 8))

    assert_equal 1, client.calls.length
    assert_equal '/app/installations/123456/access_tokens', client.calls.first[:path]
    assert_equal '12345', jwt_payload(client.calls.first[:auth_token])['iss']
  end

  def test_uses_app_id_env_var_when_app_id_is_empty
    runtime = temp_runtime
    client = fake_client

    GitHubAppToken.create_token(base_options(app_id: '', app_id_env_var: 'APP_ID',
                                             output_json: runtime[:output_json]),
                                env: env_with_private_key(runtime).merge('APP_ID' => '67890'),
                                client: client,
                                now: Time.utc(2026, 5, 8))

    assert_equal '67890', jwt_payload(client.calls.first[:auth_token])['iss']
  end

  def test_looks_up_installation_id_from_the_repository_when_not_explicit
    runtime = temp_runtime
    client = fake_client

    GitHubAppToken.create_token(base_options(installation_id: '', output_json: runtime[:output_json]),
                                env: env_with_private_key(runtime),
                                client: client,
                                now: Time.utc(2026, 5, 8))

    assert_equal 2, client.calls.length
    assert_equal '/repos/octo-org/repo-a/installation', client.calls[0][:path]
    assert_equal '/app/installations/123456/access_tokens', client.calls[1][:path]
  end

  def test_validates_permissions_json
    assert_equal({ 'contents' => 'write' }, GitHubAppToken.parse_permissions('{"contents":"write"}'))
    assert_raises(GitHubAppToken::InputError) { GitHubAppToken.parse_permissions('[]') }
    assert_raises(GitHubAppToken::InputError) { GitHubAppToken.parse_permissions('{') }
  end

  def test_rejects_repositories_and_repository_ids_together
    runtime = temp_runtime

    error = assert_raises(GitHubAppToken::InputError) do
      GitHubAppToken.create_token(base_options(repositories: 'repo-a', repository_ids: '1',
                                               output_json: runtime[:output_json]),
                                  env: env_with_private_key(runtime),
                                  client: fake_client)
    end
    assert_match(/cannot be used together/, error.message)
  end

  def test_omits_token_from_output_json_by_default_and_exports_it_through_bash_env
    runtime = temp_runtime

    GitHubAppToken.create_token(base_options(output_json: runtime[:output_json]),
                                env: env_with_private_key(runtime),
                                client: fake_client,
                                now: Time.utc(2026, 5, 8))

    output = JSON.parse(File.read(runtime[:output_json]))
    assert_nil output['token']
    assert_equal 'GH_APP_TOKEN', output['token_exported_to']
    assert_equal 123456, output['installation_id']
    assert_match(/export GH_APP_TOKEN='ghs_example_token'/, File.read(runtime[:bash_env]))
  end

  def test_redacts_private_key_and_token_values_from_errors
    message = GitHubAppToken.redact("failed with #{@private_key_pem} and ghs_example_token",
                                    [@private_key_pem, 'ghs_example_token'])

    refute_match(/BEGIN RSA PRIVATE KEY/, message)
    refute_match(/ghs_example_token/, message)
    assert_match(/\[REDACTED\]/, message)
  end

  private

  def base_options(overrides = {})
    {
      app_id: '12345',
      app_id_env_var: 'GITHUB_APP_ID',
      private_key_env_var: 'GITHUB_APP_PRIVATE_KEY',
      private_key_base64: false,
      installation_id: '123456',
      installation_id_env_var: 'GITHUB_APP_INSTALLATION_ID',
      owner: 'octo-org',
      repo: 'repo-a',
      installation_lookup: 'repo',
      repositories: '',
      repository_ids: '',
      permissions: '',
      github_api_url: 'https://api.github.com',
      github_api_version: '2026-03-10',
      token_env_var: 'GH_APP_TOKEN',
      export_env_prefix: 'GH_APP',
      output_json: '/tmp/github-app-token-result.json',
      include_token_in_output_json: false
    }.merge(overrides)
  end

  def env_with_private_key(runtime)
    {
      'BASH_ENV' => runtime[:bash_env],
      'GITHUB_APP_PRIVATE_KEY' => @private_key_pem
    }
  end

  def fake_client
    FakeGitHubClient.new(repo_installation: @repo_installation, access_token: @access_token)
  end

  def temp_runtime
    directory = Dir.mktmpdir('github-app-token-test-')
    {
      directory: directory,
      bash_env: File.join(directory, 'bash_env'),
      output_json: File.join(directory, 'result.json')
    }
  end

  def jwt_payload(token)
    payload = token.split('.')[1]
    padded = payload.ljust((payload.length + 3) / 4 * 4, '=')
    JSON.parse(Base64.urlsafe_decode64(padded))
  end
end

File: tests/create_token.test.mjs
import assert from 'node:assert/strict';
import crypto from 'node:crypto';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import test from 'node:test';

import {
  createToken,
  normalizePrivateKey,
  parsePermissions,
  redact,
} from '../src/scripts/create_token.mjs';

const repoInstallation = JSON.parse(
  fs.readFileSync(new URL('./fixtures/responses/repo-installation.json', import.meta.url), 'utf8'),
);
const accessToken = JSON.parse(
  fs.readFileSync(new URL('./fixtures/responses/access-token.json', import.meta.url), 'utf8'),
);

const { privateKey } = crypto.generateKeyPairSync('rsa', { modulusLength: 2048 });
const privateKeyPem = privateKey.export({ type: 'pkcs1', format: 'pem' });

const tempRuntime = () => {
  const directory = fs.mkdtempSync(path.join(os.tmpdir(), 'github-app-token-test-'));
  return {
    directory,
    bashEnv: path.join(directory, 'bash_env'),
    outputJson: path.join(directory, 'result.json'),
  };
};

const baseOptions = (overrides = {}) => ({
  appId: '12345',
  appIdEnvVar: 'GITHUB_APP_ID',
  privateKeyEnvVar: 'GITHUB_APP_PRIVATE_KEY',
  privateKeyBase64: false,
  installationId: '123456',
  installationIdEnvVar: 'GITHUB_APP_INSTALLATION_ID',
  owner: 'octo-org',
  repo: 'repo-a',
  installationLookup: 'repo',
  repositories: '',
  repositoryIds: '',
  permissions: '',
  githubApiUrl: 'https://api.github.com',
  githubApiVersion: '2026-03-10',
  tokenEnvVar: 'GH_APP_TOKEN',
  exportEnvPrefix: 'GH_APP',
  outputJson: '/tmp/github-app-token-result.json',
  includeTokenInOutputJson: false,
  ...overrides,
});

const jsonResponse = (body, status = 200) => ({
  ok: status >= 200 && status < 300,
  status,
  async text() {
    return JSON.stringify(body);
  },
});

const fakeFetch = () => {
  const calls = [];
  return {
    calls,
    fetchImpl: async (url, options) => {
      calls.push({
        url,
        options,
        body: options.body ? JSON.parse(options.body) : undefined,
      });

      if (url.endsWith('/repos/octo-org/repo-a/installation')) {
        return jsonResponse(repoInstallation);
      }
      if (url.endsWith('/access_tokens')) {
        return jsonResponse(accessToken);
      }
      return jsonResponse({ message: 'not found' }, 404);
    },
  };
};

const jwtPayload = (authorizationHeader) => {
  const token = authorizationHeader.replace('Bearer ', '');
  const payload = token.split('.')[1];
  return JSON.parse(Buffer.from(payload, 'base64url').toString('utf8'));
};

test('normalizes raw, escaped newline, and base64 private keys', () => {
  const expected = privateKeyPem.endsWith('\n') ? privateKeyPem : `${privateKeyPem}\n`;
  assert.equal(normalizePrivateKey(privateKeyPem), expected);
  assert.equal(normalizePrivateKey(privateKeyPem.replace(/\n/g, '\\n')), expected);
  assert.equal(normalizePrivateKey(Buffer.from(privateKeyPem).toString('base64'), { base64: true }), expected);
});

test('uses app_id parameter and skips lookup when installation_id is explicit', async () => {
  const runtime = tempRuntime();
  const { calls, fetchImpl } = fakeFetch();

  await createToken(baseOptions({ outputJson: runtime.outputJson }), {
    env: {
      BASH_ENV: runtime.bashEnv,
      GITHUB_APP_PRIVATE_KEY: privateKeyPem,
    },
    fetchImpl,
    now: new Date('2026-05-08T00:00:00Z'),
  });

  assert.equal(calls.length, 1);
  assert.match(calls[0].url, /\/app\/installations\/123456\/access_tokens$/);
  assert.equal(jwtPayload(calls[0].options.headers.Authorization).iss, '12345');
});

test('uses app_id_env_var when app_id is empty', async () => {
  const runtime = tempRuntime();
  const { calls, fetchImpl } = fakeFetch();

  await createToken(baseOptions({ appId: '', appIdEnvVar: 'APP_ID', outputJson: runtime.outputJson }), {
    env: {
      APP_ID: '67890',
      BASH_ENV: runtime.bashEnv,
      GITHUB_APP_PRIVATE_KEY: privateKeyPem,
    },
    fetchImpl,
    now: new Date('2026-05-08T00:00:00Z'),
  });

  assert.equal(jwtPayload(calls[0].options.headers.Authorization).iss, '67890');
});

test('looks up installation_id from the repository when not explicit', async () => {
  const runtime = tempRuntime();
  const { calls, fetchImpl } = fakeFetch();

  await createToken(baseOptions({ installationId: '', outputJson: runtime.outputJson }), {
    env: {
      BASH_ENV: runtime.bashEnv,
      GITHUB_APP_PRIVATE_KEY: privateKeyPem,
    },
    fetchImpl,
    now: new Date('2026-05-08T00:00:00Z'),
  });

  assert.equal(calls.length, 2);
  assert.match(calls[0].url, /\/repos\/octo-org\/repo-a\/installation$/);
  assert.match(calls[1].url, /\/app\/installations\/123456\/access_tokens$/);
});

test('validates permissions JSON', () => {
  assert.deepEqual(parsePermissions('{"contents":"write"}'), { contents: 'write' });
  assert.throws(() => parsePermissions('[]'), /JSON object/);
  assert.throws(() => parsePermissions('{'), /valid JSON object/);
});

test('rejects repositories and repository_ids together', async () => {
  const runtime = tempRuntime();
  const { fetchImpl } = fakeFetch();

  await assert.rejects(
    createToken(baseOptions({ repositories: 'repo-a', repositoryIds: '1', outputJson: runtime.outputJson }), {
      env: {
        BASH_ENV: runtime.bashEnv,
        GITHUB_APP_PRIVATE_KEY: privateKeyPem,
      },
      fetchImpl,
    }),
    /cannot be used together/,
  );
});

test('omits token from output JSON by default and exports it through BASH_ENV', async () => {
  const runtime = tempRuntime();
  const { fetchImpl } = fakeFetch();

  await createToken(baseOptions({ outputJson: runtime.outputJson }), {
    env: {
      BASH_ENV: runtime.bashEnv,
      GITHUB_APP_PRIVATE_KEY: privateKeyPem,
    },
    fetchImpl,
    now: new Date('2026-05-08T00:00:00Z'),
  });

  const output = JSON.parse(fs.readFileSync(runtime.outputJson, 'utf8'));
  assert.equal(output.token, undefined);
  assert.equal(output.token_exported_to, 'GH_APP_TOKEN');
  assert.equal(output.installation_id, 123456);
  assert.match(fs.readFileSync(runtime.bashEnv, 'utf8'), /export GH_APP_TOKEN='ghs_example_token'/);
});

test('redacts private key and token values from errors', () => {
  const message = redact(`failed with ${privateKeyPem} and ghs_example_token`, [
    privateKeyPem,
    'ghs_example_token',
  ]);

  assert.doesNotMatch(message, /BEGIN RSA PRIVATE KEY/);
  assert.doesNotMatch(message, /ghs_example_token/);
  assert.match(message, /\[REDACTED\]/);
});
